metrics/metrics.go

```diff
@@ -37,44 +41,62 @@ var (
 
 	// Used in fallback checks, e.g. to avoid sending "counters" with 0 values
 	metricTypes map[MetricID]MetricType
+
+	// OTel metric instrumentation
+	meter    = otel.Meter("go.opentelemetry.io/ebpf-profiler")
+	counters = map[MetricID]metric.Int64Counter{}
+	gauges   = map[MetricID]metric.Int64Gauge{}
 )
 
 func init() {
-	defs, err := GetDefinitions()
-	if err != nil {
-		panic("extracting definitions from metrics.json")
-	}
-
+	defs := GetDefinitions()
 	metricTypes = make(map[MetricID]MetricType, len(defs))
 	for _, md := range defs {
+		if md.Obsolete {
+			continue
+		}
 		metricTypes[md.ID] = md.Type
+		switch typ := md.Type; typ {
+		case MetricTypeCounter:
+			counter, err := meter.Int64Counter(md.Name,
+				metric.WithDescription(md.Description),
+				metric.WithUnit(md.Unit))
+			if err != nil {
+				log.Errorf("Creating Int64Counter: %v", err)
+				continue
+			}
+			counters[md.ID] = counter
+		case MetricTypeGauge:
+			gauge, err := meter.Int64Gauge(md.Name,
+				metric.WithDescription(md.Description),
+				metric.WithUnit(md.Unit))
+			if err != nil {
+				log.Errorf("Creating Int64Gauge: %v", err)
+				continue
+			}
+			gauges[md.ID] = gauge
+		default:
+			panic(fmt.Sprintf("Unknown metric type: %v", typ))
+		}
 	}
 }
 
-// reporterImpl allows swapping out the global metrics reporter.
-//
-// nil is a valid value indicating that metrics should be voided.
-var reporterImpl reporter.MetricsReporter
-
-// SetReporter sets the reporter instance used to send out metrics.
-func SetReporter(r reporter.MetricsReporter) {
-	reporterImpl = r
-}
-
-// report converts and reports collected metrics via the reporter package.
+// report converts and reports collected metrics via OTel metrics
 func report() {
-	ids := make([]uint32, nMetrics)
-	values := make([]int64, nMetrics)
-
+	ctx := context.Background()
```

Comment:
To be able to cancel the operation and not get blocked, it might be best to propagate the context to the function and us it.
```suggestion
	func report(ctx context.Context)
```

Reply:
If we add `context` it will propagate out of `report` and into every `Add/AddSlice` operation we do in the agent, which is what I was trying to avoid. Based on the initial research that I did, it didn't seem possible that a metering operation would block so that would make propagating `context` for this reason unnecessary. I'll dig some more.